debezium-server/debezium-server-eventhubs/src/main/java/io/debezium/server/eventhubs/EventHubsChangeConsumer.java

```diff
@@ -125,57 +125,74 @@ public void handleBatch(List<ChangeEvent<Object, Object>> records,
             op.setMaximumSizeInBytes(maxBatchSize);
         }
 
-        EventDataBatch batch = producer.createBatch(op);
+        for (int i = 0; i < records.size();) {
+            int added = 0;
+            int start = i;
+            LOGGER.trace("Emitting events starting from index {}", start);
+
+            EventDataBatch batch = producer.createBatch(op);
+
+            // this loop adds as many records to the batch as possible
+            for (; i < records.size();) {
+                ChangeEvent<Object, Object> record = records.get(i);
+                LOGGER.trace("Received record '{}'", record.value());
+                if (null == record.value()) {
+                    continue;
+                }
 
-        // this loop just adds records to the batch
-        for (ChangeEvent<Object, Object> record : records) {
-            LOGGER.trace("Received record '{}'", record.value());
-            if (null == record.value()) {
-                continue;
-            }
-            EventData eventData = null;
+                EventData eventData = null;
+                if (record.value() instanceof String) {
+                    eventData = new EventData((String) record.value());
+                }
+                else if (record.value() instanceof byte[]) {
+                    eventData = new EventData(getBytes(record.value()));
+                }
 
-            if (record.value() instanceof String) {
-                eventData = new EventData((String) record.value());
-            }
-            else if (record.value() instanceof byte[]) {
-                eventData = new EventData(getBytes(record.value()));
-            }
-            try {
-                if (!batch.tryAdd(eventData)) {
-                    throw new DebeziumException("Event data was too large to fit in the batch");
+                try {
+                    if (!batch.tryAdd(eventData)) {
+                        // reached the maximum allowed size for the batch
```

Comment:
Can you add a check here to make sure the eventData size is not larger than the maxBatchSize. To prevent an infinite loop.

Reply:
Shouldn't that be handled automatically by the catch of `AmqpException`?

Reply:
So I think the best way to detect the infinite-loop is when `added=0` and the `tryAdd` fails.  This way we know we've reached a situation where the event's data exceeds the maxBatchSize and we'll just throw an exception in this use case.  I believe this should ultimately be better than what it was previously.

Reply:
Ah you are right, AmqpException should take care of this. I don't think you NEED to do the added=0 check, because i don't see a case where it is triggered unless AmqpException isn't working properly.